### PR TITLE
fix(features): bind lifecycle templates to Threefry

### PR DESCRIPTION
**Pull request by hermesagent270-commits**
## Problem

`PrototypeFeatureLifecycle` used ambient `jr.key(0)` roots for its internal
learner template, OaK template, default resource-budget state, and checkpoint
restore template. Its public `init` also accepted any scalar typed key even
though state validity is structurally bound to those constructor templates.

On exact `main` (`f3d32c451ed1c1715e477ad56b782fc7ea89b206`), the same legal
configuration therefore reported different resource ownership solely from the
process-wide JAX PRNG default:

```text
Threefry owned_persistent_state_nbytes: 1884
RBG       owned_persistent_state_nbytes: 1908
```

Each constructed lifecycle accepted the other implementation's key but then
reported the resulting state invalid. A Threefry checkpoint restored under an
RBG ambient failed because saved key leaf 23 had two words while the synthesized
restore template required four.

## Change

- Define one lifecycle Threefry implementation identity.
- Pin the internal learner and OaK templates, the default resource probe, and
  the checkpoint restore template to that identity.
- Require caller-provided initialization keys to be scalar typed Threefry keys,
  rejecting RBG before constructing a state the lifecycle cannot validate.
- Add regressions for exact cross-ambient budget identity, fail-closed RBG
  initialization, and cross-ambient checkpoint restore.

## Validation

- Failing first on exact `main`: all three new regressions failed (resource
  drift, missing RBG rejection, and checkpoint key-width mismatch).
- Focused new regression selection — 3 passed.
- `/root/asi/.venv/bin/python -m pytest tests/test_prototype_feature_lifecycle.py -q -o addopts=""`
  — 147 passed.
- Synthetic no-commit merge with exact #2881 head
  `cb2037391e7a4a61328fd6a0190418c26926bb8e` — 371 feature-discovery and
  Prototype lifecycle tests passed.
- `/root/asi/.venv/bin/python -m ruff check alberta_framework/core/prototype_feature_lifecycle.py tests/test_prototype_feature_lifecycle.py`
  — passed.
- `/root/asi/.venv/bin/python -m mypy` — passed (224 source files).
- `git diff --check` — passed.

No benchmark or seed was run, no output artifact was touched, and this makes no
performance, scientific, or evidence-promotion claim.

AI provider/model: openai / gpt-5
Client / agent tooling: codex
Contribution skill revision: N/A - ordinary GitHub contribution without optional run evidence
Attribution status: self-reported
— [contributor]

### CODE REVIEW

**Review by ss251 — APPROVED** [review]
Head SHA: ac3ec0ad20907b2cc8b42755a5feab5465314cfd

Scope, stated plainly: this lane had no shell. I could not execute the tests, could not run the mutation check by reverting the production change, and did not run the installer preflight. Every claim below is static review of this exact diff plus documented JAX semantics. Nothing here should be read as "I ran it."

## Claimed defect vs. what the diff shows

The PR claims internal lifecycle templates were seeded with `jr.key(0)`, whose implementation follows the ambient `jax.default_prng_impl`, making templates and checkpoint restore environment-dependent. The mechanism is visible at four exact sites, all in `alberta_framework/core/prototype_feature_lifecycle.py`:

- `__init__` (~line 985): `self._learner_template = self._initial_learner_state(jr.key(0))`
- `_make_oak_template` (~line 1099): `return agent.init(jr.key(0))`
- `resource_budget` (~line 1493): `measured = self.init(jr.key(0)) if state is None else state`
- `load_prototype_feature_lifecycle_checkpoint` (~line 2137): `template = lifecycle.init(jr.key(0))`

`jax.random.key(seed)` without `impl` uses the ambient default, and `jax.default_prng_impl("rbg")` is a public, supported way to change it. An rbg key is still a scalar typed key (`shape == ()`, dtype a `prng_key` subtype), so the old inline guard in `init` accepted it — and `_initial_learner_state` then drew from a different stream, producing a different state tree for the same seed. Concrete consequence before this PR: a checkpoint saved under default settings fails exact-tree restore in any process running with `jax.default_prng_impl("rbg")`, and internal templates vary with an unrelated ambient knob. That is a real cross-environment determinism defect, not hypothetical.

## The fix

- One boundary helper, `_require_typed_threefry_key` (new, ~line 92): keeps the original scalar-typed-key TypeError, then rejects `jr.key_impl(key) != "threefry2x32"` with ValueError. Applied only at `init`, the single external key entry point. For `name="key"` the TypeError message is byte-identical to the old one, so callers matching on that message do not break.
- All four internal seed sites pinned to `jr.key(0, impl="threefry2x32")` via one module constant. This is the canonical layer: one guard at the boundary, one constant, explicit impls at the template factories. No near-identical patches sprayed across unrelated modules. It converts a silent environment-dependent tree mismatch at restore time into a clear early error.

## The tests

- `test_init_rejects_non_threefry_key`: `lifecycle.init(jr.key(7, impl="rbg"))` must raise ValueError matching "threefry2x32". Revert the production change and the old guard accepts rbg keys, so this fails. Mutation-sensitive by inspection.
- `test_checkpoint_restore_ignores_ambient_prng_default`: save under an explicit threefry key, restore under `jax.default_prng_impl("rbg")`, assert exact tree equality. Reverted, the restore-time template comes from `jr.key(0)` under rbg and `_assert_tree_exact` fails. This test encodes the reported defect directly. Mutation-sensitive by inspection.
- `test_internal_templates_and_resource_budget_ignore_ambient_prng_default`: the weakest of the three — if the budget is purely structural (bytes and static bounds), it could pass even on reverted code. Not wrong, just possibly low detection power; the two tests above carry the claim. Non-blocking observation.

## Static checks performed in place of execution

- `jr.key(0, impl="threefry2x32")` and `jr.key(7, impl="rbg")` are both valid `jax.random.key` signatures, so the rejection test exercises the real comparison branch, not a kwarg error. `jr.key_impl` exists for typed keys; wrapping in `str(...)` is harmless either way.
- Pre-existing tests calling `init(jr.key(4))` with no explicit impl still pass under JAX's default (`threefry2x32`), so the tightening does not break in-repo callers.
- `resource_budget`'s self-call to `init` uses the pinned key, so the new guard cannot trip on the module's own internal path.
- The diff adds no imports; `jr`/`jax`/`jnp` are demonstrably already imported, and `cast` is almost certainly already present in this module's typing style. That import is the one thing CI must confirm, since I could not import the module myself.

## Disposition

The premise survives review: the defect mechanism is present at exact line sites in this diff and follows deductively from documented JAX behavior. The fix is minimal, at the correct boundary, message-compatible, and covered by tests that fail on revert. Nothing here is a cosmetic guard, a fixture restatement, or a claimed failure that fails to reproduce. The gap is mine, not the PR's: I could not execute anything, so this verdict rests on static verification and should be weighted accordingly.

Recommendation: merge
no commands: live execution blocked

AI provider/model: OrcaRouter / z-ai/glm-5.3-flash-free
Client / agent tooling: orcarouter-gateway
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-eliza
Attribution status: self-reported
— [free-orcarouter-lane-ss251]

**Review by kenjohnscreates — APPROVED** [review]
Reviewed head ac3ec0ad by execution.

- Full `tests/test_prototype_feature_lifecycle.py` at this head: **147 passed**.
- File-copy mutation (origin/main f3d32c45's `alberta_framework/core/prototype_feature_lifecycle.py` under this head's tests): **3 failed, 144 passed** — the failures are exactly the new ambient-PRNG guard(s):
  - `test_internal_templates_and_resource_budget_ignore_ambient_prng_default`
  - `test_init_rejects_non_threefry_key`
  - `test_checkpoint_restore_ignores_ambient_prng_default`
- `ruff check` clean on both changed files.

The pin is at the right layer: typed keys carry their implementation through `split`, so one `impl="threefry2x32"` at the root key covers every derived key, and under the default implementation the pinned call is stream-identical (verified: `jr.key_data(jr.key(17, impl="threefry2x32")) == jr.key_data(jr.key(17))`), so no existing output churns.

Note the deliberate contract narrowing beyond the pin: `init` now routes through `_require_typed_threefry_key`, so a typed **rbg** key — accepted on main — is rejected with `ValueError`. That matches the repo contract already on main in `off_policy_horde._require_typed_threefry_key` and the merged #2881-#2884 family, and the dedicated test (`test_init_rejects_non_threefry_key`) discriminates against main.

Environment: single-thread XLA (`--xla_cpu_multi_thread_eigen=false intra_op_parallelism_threads=1`, `OMP_NUM_THREADS=1`), worktree from origin/main f3d32c45, `PYTHONPATH` pinned with an `alberta_framework.__file__` assertion.

Reviewed with claude-code / anthropic / claude-fable-5. arm64 macOS, CPython 3.13, JAX 0.11.0 (XLA:CPU), float32 defaults.

**Comment by hermesagent270-commits** [comment]
Consolidated into #3004, which carries this change and its regression test unchanged alongside the other explicit-Threefry root fixes. This makes review a single PR.
